### PR TITLE
Issue14725 getopt better exceptions

### PR DESCRIPTION
**Pull request by burner**
https://issues.dlang.org/show_bug.cgi?id=14725

### CODE REVIEW

**Comment by DmitryOlshansky** [comment]
LGTM

File: std/getopt.d [hunk 1]
```diff
@@ -614,6 +615,24 @@ private void getoptImpl(T...)(ref string[] args, ref configuration cfg,
     }
 }
 
+private void handleConversation(R)(string src, R* receiver,
```
Comment (schveiguy) [inline std/getopt.d]: What do you mean by "conversation"? Did you mean conversion?

Comment (burner) [inline std/getopt.d]: indeed

File: std/getopt.d [hunk 2]
```diff
@@ -614,6 +615,24 @@ private void getoptImpl(T...)(ref string[] args, ref configuration cfg,
     }
 }
 
+private void handleConversion(R)(string src, R* receiver,
+        immutable size_t idx, string option, string file = __FILE__,
+        long line = __LINE__)
```
Comment (dcarp) [inline std/getopt.d]: size_t line

File: std/getopt.d [hunk 3]
```diff
@@ -614,6 +615,24 @@ private void getoptImpl(T...)(ref string[] args, ref configuration cfg,
     }
 }
 
+private void handleConversion(R)(string src, R* receiver,
+        immutable size_t idx, string option, string file = __FILE__,
```
Comment (dcarp) [inline std/getopt.d]: immutable is unnecessary

File: std/getopt.d [hunk 4]
```diff
@@ -614,6 +615,24 @@ private void getoptImpl(T...)(ref string[] args, ref configuration cfg,
     }
 }
 
+private void handleConversion(R)(string src, R* receiver,
```
Comment (dcarp) [inline std/getopt.d]: here I would keep the order of parameters like of `handleOption`.

```D
void handleConversation(R)(string option, R receiver, string value, size_t index,
        string file = __FILE__,  size_t line = __LINE__`)
```

File: std/getopt.d [hunk 5]
```diff
@@ -695,20 +713,22 @@ private bool handleOption(R)(string option, R receiver, ref string[] args,
...
-                // string receiver
-                *receiver = to!(typeof(*receiver))(val);
+                if (incremental)
+                {
+                    ++*receiver;
+                }
+                else
+                {
+                    handleConversion(option, val, receiver, i);
+                }
```
Comment (quickfur) [inline std/getopt.d]: While I agree that the original code is hard to read, expanding it into 3-line braced blocks seems excessive. What about just leaving out those braces?

File: std/getopt.d [hunk 6]
```diff
@@ -1482,3 +1515,30 @@ unittest
         ~ "          This help information.\n";
     assert(wanted == helpMsg, helpMsg ~ wanted);
 }
+
+unittest
+{
+    import std.string : indexOf;
+
+    enum ENUM {
```
Comment (quickfur) [inline std/getopt.d]: This must be the only place I've ever seen in Phobos that uses ALL_CAPS identifiers. Could we use `Enum` instead? (The precedent for that is `std.typecons.BitFlags`, which also uses `Enum` in the unittest.)

File: std/getopt.d (hunk 5 above)
Comment (quickfur) [inline std/getopt.d]: I.e., leave the if-block and else-block as separate lines, but just indent them without using braces.

**Comment by schveiguy** [comment]
LGTM, @DmitryOlshansky I'll let you pull, if you still think it looks good after all the changes.

**Comment by DmitryOlshansky** [comment]
Let's wait a bit on this. I have a proposal brewing that may solve this and many exception-hierarchy issues.

**Comment by burner** [comment]
could you post a reference?

**Comment by burner** [comment]
I'm on holiday for a good 2 weeks, I guess we can wait some more

**Comment by DmitryOlshansky** [comment]
> I'm on holiday for a good 2 weeks, I guess we can wait some more

And I was about to post that I'm on a cellphone for at least a week. Let's get back to this in ~ 10 days.

**Comment by burner** [comment]
@DmitryOlshansky ping

**Comment by DmitryOlshansky** [comment]
@burner got busy with real life. Ping me back in a couple of days, I'll prepare the proposal

**Comment by quickfur** [comment]
ping
Any news on this?

**Comment by burner** [comment]
well, the bugfix is still good.
But @DmitryOlshansky said he had some magic pill. But he didn't say what is was or when he will present it.

**Comment by DmitryOlshansky** [comment]
> But @DmitryOlshansky said he had some magic pill. But he didn't say what is was or when he will present it.

Simply put the current fix silently changes exception type. Now users catching ConvException would be surprized to see that they no longer catch the right condition. As for magic - yes I have it, have patience my friends.

**Comment by DmitryOlshansky** [comment]
> ping
> Any news on this?

To start on this as I think I'll wait forever to write a DIP.

What we should have here instead of this "fix" is throw an exception that is *both* ConversionException (semantics) and GetOptExcpetion (origin). I believe that I found a way to frutifully combine 2 exception hierarchies where one of them is going to be an interface hierarchy. Now my proposal wouldn't stop on just getting hierarchies right we should cover:
1. Easy creation of exceptions :
```d
alias MyParseException = StandardException!(ParseException, my.module);
alias MyLogicException = StandardException!(LogicException, my.module);
```

2.  Ease of throwing and creating exceptions:
```d
// writef style formatting
throw exceptionf!MyParseException("This is lazily formatted with %d a number", 42);
// write style formatting
throw exception!MyParseException("This is lazily formatted with ", 42, " a number");
```
3. Catching can be done by any of the two: semantic type as in ParseException or by origin (ideally) as `ExceptionOrigin!(module.name)` or `ExceptionOrigin!(package.name)`. Then:
```d
try{ ... }
catch(ParseException ){ // may be a parse error, possibly in SQL statement
} 
catch(OriginExcpetion!(arsd.mysql)) { // some exception from Mysql package as a whole
}
 ```

I believe it's this blend of origin and semantics is the way to go, it's novel I admit but it enables excellent idioms and seems to scale way beyond other hierarchies proposed.

**Comment by DmitryOlshansky** [comment]
> To start on this as I think I'll wait forever to write a DIP.

The missing piece is proof of concept and that may take a bit of time

**Comment by quickfur** [comment]
I remember you posted about this in the forum, but there wasn't very much response.

Personally I think it's a great idea.

**Comment by burner** [comment]
@DmitryOlshansky ping?

**Comment by JackStouffer** [comment]
@DmitryOlshansky If this PR is fixing a genuine problem, then why not pull this instead of waiting for something that might still take a long time. I believe you when you say you have a better solution, but this solution exists now, why not just override this with your changes when they're ready?

**Comment by DmitryOlshansky** [comment]
This just changes one exception for another - a breaking change that helps users that didn't expect ConvError but breakscode of others that expect it. Closing one ticket foranew one

**Comment by JakobOvrum** [comment]
Yeah, changing the type of a thrown exception breaks code, so let's do it as few times as possible. I agree with @DmitryOlshansky; I think this patch is better applied after his work, when it will be backwards compatible.

**Comment by burner** [comment]
@DmitryOlshansky any news on you DIP

**Comment by DmitryOlshansky** [comment]
> @DmitryOlshansky any news on you DIP

Not yet, though I intend to process my D-language related backlog soonish.

**Comment by wilzbach** [comment]
> @DmitryOlshansky any news on you[r] DIP

ping @DmitryOlshansky - ´I like your idea of having multiple exceptions. Any way we can help you?

**Comment by DmitryOlshansky** [comment]
> ´I like your idea of having multiple exceptions. Any way we can help you?

Well so far an attempt to convert Throwable to interface instead of a class exploded in my face at run-time. I need it done and it's the only real obstacle. I will publish the fork of druntime in case there is any interest

**Comment by JackStouffer** [comment]
If this is waiting on @DmitryOlshansky's work, then IMO this should be closed, as this will require a rewrite anyway. Right now this is just eating auto-tester time.

**Comment by burner** [comment]
agreeded

**Pull request closed without merging**